src/main/kotlin/org/jitsi/nlj/transform/node/Node.kt

```diff
@@ -109,58 +96,200 @@ abstract class Node(
-    }
+    protected fun doneProcessing(packetInfo: PacketInfo?) {
+        val processingDuration = System.nanoTime() - startTime
+        totalProcessingDuration += processingDuration
 
-    private fun onEntry(incomingPackets: List<PacketInfo>) {
-        startTime = System.nanoTime()
-        if (firstPacketTime == -1L) {
-            firstPacketTime = startTime
-        }
-        incomingPackets.forEach {
-            numBytes += it.packet.sizeBytes
-            it.addEvent(nodeEntryString)
-        }
-        lastPacketTime = System.nanoTime()
-        numInputPackets += incomingPackets.size
+        packetInfo?: numOutputPackets++
+        packetInfo?.addEvent(nodeExitString)
     }
 
-    private fun onExit(outPackets: List<PacketInfo>) {
+    /**
+     * Should be called by sub classes when they finish processing of the input packet, but before they call into any
+     * other nodes, so that [Node] can keep track of its statistics.
+     */
+    protected fun doneProcessing(packetInfos: List<PacketInfo>) {
         val processingDuration = System.nanoTime() - startTime
         totalProcessingDuration += processingDuration
-        outPackets.forEach {
+
+        numOutputPackets += packetInfos.size
+        packetInfos.forEach {
             it.addEvent(nodeExitString)
         }
     }
 }
 
+/**
+ * A [Node] which transforms a single packet, possibly dropping it (by returning null).
+ */
+abstract class TransformerNode(
+    name: String
+) : StatsKeepingNode(name) {
+
+    protected abstract fun transform(packetInfo: PacketInfo): PacketInfo?
+
+    override fun doProcessPacket(packetInfo: PacketInfo) {
+        val transformedPacket = transform(packetInfo)
+        doneProcessing(transformedPacket)
+        next(packetInfo)
+    }
+}
+
+/**
+ * A [Node] which drops some of the packets (the ones which are not accepted).
+ */
+abstract class FilterNode(
+    name: String
+) : TransformerNode(name) {
+
+    protected abstract fun accept(packetInfo: PacketInfo): Boolean
+
+    override fun transform(packetInfo: PacketInfo): PacketInfo? {
+        return if (accept(packetInfo)) packetInfo else null
+    }
+}
+
+typealias PacketInfoPredicate = Predicate<PacketInfo>
+abstract class PredicateFilterNode(
+    name: String,
+    val predicate: PacketInfoPredicate): FilterNode(name) {
+    override fun accept(packetInfo: PacketInfo): Boolean {
+        return predicate.test(packetInfo)
+    }
+}
+
+/**
+ * A [Node] which observes packets, but makes no modifications.
+ */
+abstract class ObserverNode(
+    name: String
+) : TransformerNode(name) {
+
+    protected abstract fun observe(packetInfo: PacketInfo)
+
+    override fun transform(packetInfo: PacketInfo): PacketInfo? {
+        observe(packetInfo)
+        return packetInfo
+    }
+}
+
+/**
+ * A node which consumes all packets (i.e. does something with them, but does not forward them to another node).
+ */
+abstract class ConsumerNode(
+    name: String
+) : TransformerNode(name) {
```

Comment:
i wonder if we should also have any attempt to 'attach' to a `ConsumerNode` throw (like demuxer node does)

Reply:
Sounds good, I did it. Note that I left detach as a no-op, because it's no harm and the node visitor might use it .